job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/DslScriptLoader.groovy

```diff
@@ -12,79 +12,114 @@ import java.util.logging.Logger
  * Runs provided DSL scripts via an external {@link JobManagement}.
  */
 class DslScriptLoader {
+
     private static final Logger LOGGER = Logger.getLogger(DslScriptLoader.name)
     private static final Comparator<? super Item> ITEM_COMPARATOR = new ItemProcessingOrderComparator()
 
-    private static GeneratedItems runDslEngineForParent(ScriptRequest scriptRequest,
-                                                        JobManagement jobManagement) throws IOException {
-        PrintStream logger = jobManagement.outputStream
+    private final JobManagement jobManagement
+    private final PrintStream logger
+
+    /**
+     * For testing a string directly.
+     */
+    static GeneratedItems runDslEngine(String scriptBody, JobManagement jobManagement) throws IOException {
+        ScriptRequest scriptRequest = new ScriptRequest(null, scriptBody, new File('.').toURI().toURL())
+        runDslEngine(scriptRequest, jobManagement)
+    }
+
+    static GeneratedItems runDslEngine(ScriptRequest scriptRequest, JobManagement jobManagement) throws IOException {
+        DslScriptLoader loader = new DslScriptLoader(jobManagement)
+        loader.runScripts([scriptRequest])
+    }
+
+    DslScriptLoader(JobManagement jobManagement) {
+        this.jobManagement = jobManagement
+        this.logger = jobManagement.outputStream
+    }
 
+    GeneratedItems runScripts(Collection<ScriptRequest> scriptRequests) throws IOException {
         ClassLoader parentClassLoader = DslScriptLoader.classLoader
         CompilerConfiguration config = createCompilerConfiguration(jobManagement)
 
         // Otherwise baseScript won't take effect
         GroovyClassLoader groovyClassLoader = new GroovyClassLoader(parentClassLoader, config)
+
         try {
-            GroovyScriptEngine engine = new GroovyScriptEngine(scriptRequest.urlRoots, groovyClassLoader)
+            runScriptsWithClassLoader(scriptRequests, groovyClassLoader, config)
+        } finally {
+            if (groovyClassLoader instanceof Closeable) {
+                ((Closeable) groovyClassLoader).close()
+            }
+        }
+    }
+
+    private GeneratedItems runScriptsWithClassLoader(Collection<ScriptRequest> scriptRequests,
+                                                     GroovyClassLoader groovyClassLoader,
+                                                     CompilerConfiguration config) {
+        GeneratedItems generatedItems = new GeneratedItems()
+
+        // group requests that share the same classpath
+        scriptRequests.groupBy { it.urlRoots*.toString().sort() }.values().each { List<ScriptRequest> requestSet ->
```

Comment:
Hm, this introduces another incompatibility. Users may rely on the script execution order, e.g. they have one script generating some folders and then other scripts in sub directories generating the job in these folders. If we change the execution order, it's no longer guaranteed that the folders are generated before the jobs. See [JENKINS-33081](https://issues.jenkins-ci.org/browse/JENKINS-33081).

How about iterating over the script requests and creating a new script engine when the classpath changes? When using `jobs/*.groovy` this would still be efficient, but when using weird script locations it would still be compatible.